### DISCUSSION TITLE
Move-PnPFile from a site to personal onedrive site

### DISCUSSION THREAD

**Discussion by omarmallat** [pnp/powershell#667] [Q&A]
I'm trying to move some files from SharePoint Online Site to a user OneDrive for business. This feature is available on SharePoint online web interface, but it will be easier to do it using PowerShell codes.
`Move-PnPFile -SourceUrl "/sites/TestSite/Shared Documents/General/Agreement.doc" -TargetUrl "/personal/testuser_testdomain_com/Documents" -AllowSchemaMismatch -AllowSmallerVersionLimitOnDestination`

it failed saying `The remote server returned an error: (404) Not Found`.

I feel that the reason behind this failure the base URL that I used to connect. 

1. For SharePoint Site, URL start with https://Test.sharepoint.com/sites/TestSite 
2. For OneDrive,  The URL start with https://Test-my.sharepoint.com/personal/testuser_testdomain_com/

**Comment by LeonArmston** [comment]
Hi 

This looks very similar to this discussion question: https://github.com/pnp/powershell/discussions/620#discussioncomment-717269. 

You could use the concept in my post linked above in reverse to add the files to stream objects to migrate from SP to OneDrive.

**Reply by omarmallat** [reply]
then how they are offering this feature on the web? for sure there is a way

**Reply by LeonArmston** [reply]
Yes there would be a way to do this on the web as there is probably code in the web to pass the authentication used with OneDrive to SharePoint. 

I suspect this is not happening with PnP powershell at the moment but maybe possible with a future update to these cmdlets to handle passing auth from OneDrive to SharePoint. Otherwise the workaround at the moment is to use multiple cmdlets to do the authentication to both sites.

**Comment by korah91** [comment]
Hello everyone,

I have the same problem and have found no solution...
I am trying to move all the files from a OneDrive user to a Sharepoint site in the same tenant. I have found other cases such as #620 and the case https://github.com/pnp/PnP-PowerShell/issues/1719 (I suppose this is an old repo for an older version of PnP).
In https://learn.microsoft.com/en-us/answers/questions/1186568/copy-file-from-onedrive-to-sharepoint-with-powersh someone answers that you have to use Connect-PnPOnline twice, each with one of the addresses (OneDrive user and Sharepoint Site), however, it gives me 403 Forbidden error even though I have all the permissions (Global Admin and Site Administrator of that site).

I have also tried using the same command as #3215 , but it gives me the 
`The system cannot find the file specified. (Exception from HRESULT: 0x80070002)\"."}}}` 
error in the Target Url. It is supposed to be fixed #3245 , but it doesn't appear like that to me.

**What is the version of the Cmdlet module you are running?**
`(you can retrieve this by executing Get-Module -Name "PnP.PowerShell" -ListAvailable)`
2.4.0

Thanks for the help :)

**Comment by omarmallat** [comment]
It works for me, but from SharePoint to OneDrive. The main command to use is `Copy-PnPFile `

first make sure you connect to both sites, the sharepoint and the onedrive, and save these connnections
```
    $TargetConnection = Connect-PnPOnline -URL $TargetSiteURL -Interactive -ReturnConnection
    $SourceConnection = Connect-PnPOnline -URL $SourceSiteURL -Interactive -ReturnConnection

```
then use the following command:
`Copy-PnPFile -Connection $SourceConnection -SourceUrl $file.ServerRelativeUrl -TargetUrl $TargetFolderPath -Force`


the $file.ServerRelativeUrl looks like: Shared Documents/2023/Folder1

the $TargetFolderPath looks like: https://mydomain-my.sharepoint.com/personal/user_mydomain_com/Documents/Projects/Operations/X1/X2

**Reply by korah91** [reply]
Thanks!!!
I think the problem was that I was not making the connections to both URLs.
However, I dont understand why you save the target connection if you only use the source connection in the Copy-PnPFile cmdlet
Thank you for your help! 
I think the [Copy-PnPFile documentation](https://pnp.github.io/powershell/cmdlets/Copy-PnPFile.html#description) is wrong, as it states that _It is currently not possible to copy files between a OneDrive for Business site to SharePoint or vice versa._. And I have been able to do that

**Reply by omarmallat** [reply]
There is no actual reason behind saving the target connection. The whole code was a result of many updates and testing over years. it was only few months ago when the pnp.powershell was updated to enable copying from SharePoint to OneDrive.
